### PR TITLE
Support kerberos authentication for Kudu connector

### LINKED ISSUE #1237: Support Kerberos authentication for Kudu Connector
When I use presto to connect kudu tables,I can show tables and describe kudu table's structure,But I can not select data from kudu table,there is error as following:
Query 20190802_074604_00006_2g8pc failed: org.apache.kudu.client.NonRecoverableException: Couldn't find a valid master in (192.168.1.1:7051,192.168.1.2:7051,192.168.1.3:7051). Exceptions received: [org.apache.kudu.client.NonRecoverableException: server requires authentication, but client does not have Kerberos credentials (tgt). Authentication tokens were not used because no token is available, org.apache.kudu.client.NonRecoverableException: server requires authentication, but client does not have Kerberos credentials (tgt). Authentication tokens were not used because no token is available, org.apache.kudu.client.NonRecoverableException: server requires authentication, but client does not have Kerberos credentials (tgt). Authentication tokens were not used because no token is available]

Anyone has occured this problem?

### PR DESCRIPTION
**Pull request by liyubin117**
Add the following configuration to etc/catalog/kudu.properties to enable kerberos authentication:

>        ## Whether to enable kerberos authentication, default is false
>        kudu.kerberos-auth.enabled=true
> 
>        # whether to output kerberos debug information, default is false
>        kudu.kerberos-auth.debug.enabled=true
> 
>        # The Kerberos principal that Presto will use when connecting to Kudu
>        kudu.kerberos-auth.principal=xxx
> 
>        # Kudu client keytab location
>        kudu.kerberos-auth.keytab=xxx.keytab

```
Kudu Changes

support kerberos authentication
```

Fixes #1237

### CODE REVIEW

File: presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java [hunk 1]
```diff
@@ -81,15 +82,18 @@
     public static final String DEFAULT_SCHEMA = "default";
     private final KuduClient client;
     private final SchemaEmulation schemaEmulation;
+    private final boolean kerberosAuthEnabled;
 
-    public KuduClientSession(KuduClient client, SchemaEmulation schemaEmulation)
+    public KuduClientSession(KuduClient client, SchemaEmulation schemaEmulation, boolean kerberosAuthEnabled)
```
Comment (Praveen2112) [inline presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java:87]: Can we have it as a wrap of top of KuduClientSession ?

File: presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduModule.java [hunk 2]
```diff
@@ -83,14 +83,14 @@ KuduClientSession createKuduClientSession(KuduClientConfig config)
     {
         requireNonNull(config, "config is null");
 
-        KuduClient.KuduClientBuilder builder = new KuduClient.KuduClientBuilder(config.getMasterAddresses());
-        builder.defaultAdminOperationTimeoutMs(config.getDefaultAdminOperationTimeout().toMillis());
-        builder.defaultOperationTimeoutMs(config.getDefaultOperationTimeout().toMillis());
-        builder.defaultSocketReadTimeoutMs(config.getDefaultSocketReadTimeout().toMillis());
-        if (config.isDisableStatistics()) {
-            builder.disableStatistics();
+        KuduClient client;
```
Comment (Praveen2112) [inline presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduModule.java:86]: Can we have separate module for the kerberos based KuduClient ?

File: presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduUtil.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+     */
+    static void initKerberosENV(String principal, String keytab, boolean debugEnabled)
+    {
+        try {
+            Configuration conf = new Configuration(false);
+            conf.set("hadoop.security.authentication", "kerberos");
+            if (debugEnabled) {
+                System.setProperty("sun.security.krb5.debug", "true");
+            }
+            UserGroupInformation.setConfiguration(conf);
```
Comment (Praveen2112) [inline presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduUtil.java:44]: Do we need to add hadoop dependencies ? How about reusing `KerberosAuthentication` in presto plugiin toolkit ?

File: presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientConfig.java [hunk 4]
```diff
@@ -138,4 +142,52 @@ public KuduClientConfig setSchemaEmulationEnabled(boolean enabled)
...
+        this.kerberosAuthEnabled = enabled;
+        return this;
+    }
+
+    public String getKerberosPrincipal()
+    {
+        return kerberosPrincipal;
+    }
+
+    @Config("kudu.kerberos-auth.principal")
```
Comment (kokosing) [inline presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientConfig.java:163]: Kerberos related properties should not be allowed to be configures if `kudu.authentication.type=NONE`. Please see usages how it typically achieved by look at usages of `io.airlift.configuration.ConditionalModule#installModuleIf`.

**Comment by liyubin117** [comment]
@kokosing @Praveen2112 Thanks for your review and suggestions!
I will complete the pull request as soon as possible, ins spite of a very urgent and unexpected internal deadline may breaks my later schedule, I will do my best

**Comment by sumit-gupta-sgt** [comment]
Any expected ETA for this functionality ?

**Comment by grantatspothero** [comment]
@sumit-gupta-sgt This feature was just released in Trino 372: 
https://trino.io/docs/current/release/release-372.html#kudu-connector

**Comment by hashhar** [comment]
Closing this since it is now obsolete and was merged in https://github.com/trinodb/trino/pull/10953

**Pull request closed without merging**